### PR TITLE
Show "Publish: Immediately" as date for new drafts without a date

### PR DESCRIPTION
**Pull request by earnjam**
## Description
This fixes the issue with the post date label discovered on #7195. 

Need some feedback/discussion on the REST API portion. See [my comment on #7195](https://github.com/WordPress/gutenberg/issues/7195#issuecomment-409916770) for reasons on needing the modification. It's a bit hackish to use a `null` value for `date_gmt`, but that's the way that WordPress has traditionally handled it, so it's just a continuation of that.

## How has this been tested?
Create a new post and look at the date label

## Screenshots 
Look at the publish panel on the side. This is immediately on creation of a new post right now. It sets the date the draft was first created. This is not the date that will be used when publishing. 
screen shot 2018-08-02 at 11 06 58 am

Fixed version. When no date has been set, it shows "Immediately" instead. 
screen shot 2018-08-02 at 11 07 43 am

## Types of changes
Filters the REST API response to check the post for a `post_date_gmt` value of  `0000-00-00 00:00:00` and sets `date_gmt` value to `null` if so
Bug fix on `PostScheduleLabel` component

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by earnjam** [comment]
I'm having second thoughts about this method, as modifying the response here might break back compatibility. 

An alternative proposed to core is a `date_floating` flag.

The other possibility is to check if the `date` and `modified` values are identical on a `draft` or `auto-draft`. They should likely only be identical when a true post date hasn't been set yet, but that feels like a potentially invalid assumption (though with very low risk if wrong)

**Comment by earnjam** [comment]
Updated to use a `date_floating` flag on the REST API response based on discussions on https://core.trac.wordpress.org/ticket/39953

**Review by desrosj — CHANGES_REQUESTED** [review]
I think the addition of flag makes sense over filtering the response here. Few points:

- There is a closure being used that should become a regular function definition.
- Not sure that `get_post()->post_date_gmt` will _always_ return a proper `WP_Post` object.

**Comment by earnjam** [comment]
Removed the closure, wasn't thinking there. 🙈 

The `get_callback` function can take a parameter that is an array of values of the post in the REST API response. I used that to pass an ID to the `get_post()` call. That seems safer because if there isn't a post with that ID, then it would have erred out before it got to that point. 

## Quick Summary
Posts that have a status of `draft` or `auto-draft` and do not have a date explicitly set by the user, have a date that is considered floating. Essentially this means a true post date hasn't been set, and will either be set automatically when the post is published or can be set manually by the user. When the date is floating, we should be showing "Publish: Immediately" instead of a date that may not be accurate once published. 

WordPress core traditionally uses a saved value of `0000-00-00 00:00:00` for `post_date_gmt` to determine if the post has a floating date. 

About 18 months ago, we added a shim to force the REST API to return a value for the `date_gmt` because its easier to work with than the localized `date`. See https://core.trac.wordpress.org/changeset/40108

This essentially means there is no way to know for certain that a post has a floating date. Based on some discussions in [trac ticket #39953](https://core.trac.wordpress.org/ticket/39953) it seems like adding a `date_floating` flag is the best solution for backwards compatibility. That's what I've done here.

**Comment by earnjam** [comment]
I'm not sure why Travis won't update here. I manually re-ran the build and all checks passed: https://travis-ci.org/WordPress/gutenberg/builds/412776921

**Comment by earnjam** [comment]
@danielbachhuber When you have a minute, can I get your thoughts on this approach since it will require a change to the REST API? See [my last comment above](https://github.com/WordPress/gutenberg/pull/8395#issuecomment-410825440) for a quick summary and related links to trac.

**Comment by danielbachhuber** [comment]
> When you have a minute, can I get your thoughts on this approach

Ugh, what an awful issue. I've read the Trac ticket and will give it some thought. I think @joehoyle might be at WC Pub contributor day on Friday too.

**Comment by earnjam** [comment]
Thanks. I don't like the solution for sure, but I'm struggling to come up with a better alternative.

File: lib/rest-api.php [hunk 1]
```diff
@@ -49,6 +49,34 @@ function gutenberg_register_rest_routes() {
...
+/**
+ * Register a new field on the REST API posts response for floating dates
+ *
+ * @see  https://core.trac.wordpress.org/ticket/39953
+ *
+ * TODO: This is a temporary solution until a date_floating flag can be added to core
+ */
+function gutenberg_add_date_floating_flag() {
+	$post_types = get_post_types( array( 'show_in_rest' => true ), 'names' );
+	register_rest_field( $post_types, 'date_floating', array( 'get_callback' => 'gutenberg_get_date_floating_flag' ) );
```
Comment (danielbachhuber) [inline lib/rest-api.php]: Should this be flagged as a read-only field?

**Review by danielbachhuber — CHANGES_REQUESTED** [review]
I don't love the `date_floating` attribute name, but I don't see a way around it at this point.

Could we get PHPUnit tests added for the behavior of `date_floating`, so we have explicit usage described?

**Comment by kadamwhite** [comment]
Based on the most recent conversation in https://core.trac.wordpress.org/ticket/39953 I would prefer we move forward with the solution of inferring this flag when the modified date is the same as the post date, and update the `PostScheduleLabel` component accordingly. I have proposed a small patch for the Gutenberg side of things in that ticket, which I will prepare as a separate PR.

**Comment by kadamwhite** [comment]
I propose #9967 as an alternative approach which should supersede this PR.

**Comment by earnjam** [comment]
Closing this in favor of #9967

**Pull request closed without merging**